src/mame/drivers/mk98.cpp

```diff
@@ -0,0 +1,488 @@
+
+	m_kclk = true;
+	m_kbit = 0;
+	m_scancode = 0;
+	m_kbdflag = 0;
+}
+
+uint8_t mk98_state::serial_r(offs_t offset)
+{
+	LOGDBG("aux: read  == %02X\n", m_crtc[0]);
+	return m_crtc[0] | (m_crtc[0] & 1) << 7;
+}
+
+void mk98_state::serial_w(offs_t offset, uint8_t data)
+{
+	LOGDBG("aux: write <= %02X\n", data);
+	m_crtc[0] = data & 0xe7;
+}
+
+/* keyboard HLE -- adapted from pt68k4.cpp */
+
+uint8_t mk98_state::keyboard_r(offs_t offset)
+{
+	if (offset == 0)
+	{
+		LOGKBD("kbd: read  %02X == %02X\n", offset + 0x60, m_scancode);
+		m_pic8259->ir1_w(CLEAR_LINE);
+		return m_scancode;
+	}
+	else
+		return 0;
+}
+
+void mk98_state::keyboard_w(offs_t offset, uint8_t data)
+{
+	LOGKBD("kbd: write %02X <= %02X\n", offset + 0x60, data);
+	m_pic8259->ir1_w(CLEAR_LINE);
+}
+
+WRITE_LINE_MEMBER(mk98_state::keyboard_clock_w)
+{
+	LOGKBD("kbd: KCLK: %d kbit: %d\n", state ? 1 : 0, m_kbit);
+
+	if ((state == ASSERT_LINE) && (!m_kclk))
+	{
+		if (m_kbit >= 1 && m_kbit <= 8)
+		{
+			m_scancode >>= 1;
+			m_scancode |= m_kdata;
+		}
+
+		// stop bit?
+		if (m_kbit == 9)
+		{
+			m_scancode >>= 1;
+			m_scancode |= m_kdata;
+			// arrow keys
+			switch (m_scancode)
+			{
+				case 0x48: m_scancode = 0x3c; break;
+				case 0x50: m_scancode = 0x3e; break;
+				case 0x4b: m_scancode = 0x3d; break;
+				case 0x4d: m_scancode = 0x3f; break;
+				case 0x53: m_scancode = 0x3b; break;
+			}
+			LOGKBD("kbd: scancode %02x\n", m_scancode);
+			m_kbit = 0;
+			m_pic8259->ir1_w(ASSERT_LINE);
+		}
+		else
+		{
+			m_kbit++;
+		}
+	}
+
+	m_kclk = (state == ASSERT_LINE) ? true : false;
+}
+
+WRITE_LINE_MEMBER(mk98_state::keyboard_data_w)
+{
+	LOGKBD("kbd: KDATA: %d\n", state ? 1 : 0);
+	m_kdata = (state == ASSERT_LINE) ? 0x80 : 0x00;
+}
+
+/* video HLE */
+
+uint8_t mk98_state::video_register_r()
+{
+	uint8_t ret = 0;
+
+	switch (m_register_address_latch)
+	{
+	case 0x0e:
+		ret = (m_cursor_addr >> 8) & 0xff;
+		break;
+
+	case 0x0f:
+		ret = (m_cursor_addr >> 0) & 0xff;
+		break;
+
+	/* all other registers are write only and return 0 */
+	default:
+		break;
+	}
+
+	return ret;
+}
+
+void mk98_state::video_register_w(uint8_t data)
+{
+	switch (m_register_address_latch)
+	{
+	case 1:
+		m_graphics_mode = BIT(data, 4);
+		break;
+
+	case 0x0a:
+		m_cursor_start_ras = data & 0x7f;
+		break;
+
+	case 0x0c:
+		m_disp_start_addr = ((data & 0x3f) << 8) | (m_disp_start_addr & 0x00ff);
+		break;
+
+	case 0x0d:
+		m_disp_start_addr = ((data & 0xff) << 0) | (m_disp_start_addr & 0xff00);
+		break;
+
+	case 0x0e:
+		m_cursor_addr = ((data & 0x3f) << 8) | (m_cursor_addr & 0x00ff);
+		break;
+
+	case 0x0f:
+		m_cursor_addr = ((data & 0xff) << 0) | (m_cursor_addr & 0xff00);
+		break;
+	}
+}
+
+void mk98_state::video_address_w(uint8_t data)
+{
+	m_register_address_latch = data & 0x3f;
+}
+
+uint8_t mk98_state::video_r(offs_t offset)
+{
+	int data = 0xff;
+
+	switch (offset)
+	{
+	case 0:
+		return keyboard_r(0);
+
+	case 5:
+		data = video_register_r();
+		break;
+	}
+
+	return data;
+}
```

Comment:
You’re not being consistent with whether you want to do a single return, or return from each case.  Either make the video register read like the keyboard read and get rid of `data`, or the other way around.

Out of curiosity, why go with a function to handle a range when registers only exist at specific addresses?  Would it make more sense to just map the registers that exist a the relevant addresses?

Reply:
Side effect of copy-pasting from a similar driver (hp95lx).   No idea if there are or aren't other registers in the range, thus the switch().

Reply:
If it tries to access other unknown things in the range, you’d get an unmapped access log message if you mapped the handlers directly rather than mapping across the range and using a switch.